### PR TITLE
added support for the ZRIB V2.0 control board think it's finished

### PR DESCRIPTION
**Pull request by FATALHERMIT**
theres still a few errors when compiling the lcd files something about uint_8, usec but it doesnt seem to be affecting anything

### CODE REVIEW

**Comment by FATALHERMIT** [comment]
sorry i had forgotten to set the fan pins

**Comment by FATALHERMIT** [comment]
the ramps pins seems to be overriding my settings, im working on it

File: Marlin/boards.h [hunk 1]
```diff
@@ -1,100 +1,100 @@
-/**
```
Comment (thinkyhead) [inline Marlin/boards.h]: Pleas use UNIX line endings.

File: Marlin/pins_ZRIB_V20.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
+  #include "pins_MKS_13.h"
+#include "pins_RAMPS.h"
```
Comment (thinkyhead) [inline Marlin/pins_ZRIB_V20.h]: This include is not needed (redundant).

File: Marlin/pins_ZRIB_V20.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+
+  #undef FAN_PIN
+  #undef E0_AUTO_FAN_PIN
+  #undef E1_AUTO_FAN_PIN
+  #undef E2_AUTO_FAN_PIN
+  #define FAN_PIN        ZRIB_V20_D6_PIN 
+  #define E0_AUTO_FAN_PIN    ZRIB_V20_D9_PIN 
+  #define E1_AUTO_FAN_PIN    ZRIB_V20_D9_PIN
+  #define E2_AUTO_FAN_PIN    ZRIB_V20_D9_PIN
```
Comment (thinkyhead) [inline Marlin/pins_ZRIB_V20.h]: You should override these instead:
```cpp
#undef ORIG_E0_AUTO_FAN_PIN
#undef ORIG_E1_AUTO_FAN_PIN
#undef ORIG_E2_AUTO_FAN_PIN
#undef ORIG_E3_AUTO_FAN_PIN
#define ORIG_E0_AUTO_FAN_PIN ZRIB_V20_D9_PIN
#define ORIG_E1_AUTO_FAN_PIN ZRIB_V20_D9_PIN
#define ORIG_E2_AUTO_FAN_PIN ZRIB_V20_D9_PIN
#define ORIG_E3_AUTO_FAN_PIN ZRIB_V20_D9_PIN
```

File: Marlin/pins_ZRIB_V20.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  #define ZRIB_V20_D9_PIN 9
+  #endif
+
+
+
+  #undef FAN_PIN
+  #undef E0_AUTO_FAN_PIN
+  #undef E1_AUTO_FAN_PIN
+  #undef E2_AUTO_FAN_PIN
+  #define FAN_PIN        ZRIB_V20_D6_PIN
```
Comment (thinkyhead) [inline Marlin/pins_ZRIB_V20.h]: To override all general use of the usual `FAN_PIN` (D9) you should use this line instead, before the `#include pins_MKS_13.h` line:
```cpp
#define RAMPS_D9_PIN ZRIB_V20_D6_PIN
```
This allows the pin to be reassigned to a heater or other use according to the configuration.

File: Marlin/pins_ZRIB_V20.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
+  #include "pins_MKS_13.h"
+#include "pins_RAMPS.h"
+  #undef  SERVO1_PIN 6
```
Comment (thinkyhead) [inline Marlin/pins_ZRIB_V20.h]: `#undef` doesn't take a value. If you mean to redefine it, you should use both `#undef` and `#define`.

File: Marlin/pins_ZRIB_V20.h [hunk 6]
```diff
@@ -0,0 +1,20 @@
+  #include "pins_MKS_13.h"
+#include "pins_RAMPS.h"
+  #undef  SERVO1_PIN 6
+  #ifndef ZRIB_V20_D6_PIN
+  #define ZRIB_V20_D6_PIN 6
```
Comment (thinkyhead) [inline Marlin/pins_ZRIB_V20.h]: Two spaces indentation here, and on `ZRIB_V20_D9_PIN` below.

File: Marlin/pins.h [hunk 7]
```diff
@@ -1,554 +1,556 @@
-/**
```
Comment (thinkyhead) [inline Marlin/pins.h]: Please fix the line endings in this file. You must use UNIX line endings.

**Review by thinkyhead — CHANGES_REQUESTED** [review]
Please make the requested changes, and then we'll review at the boards and pins files.

File: Marlin/pins.h (hunk 7 above)
Comment (FATALHERMIT) [reply Marlin/pins.h]: understood

File: Marlin/pins_ZRIB_V20.h [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#undef SERVO1_PIN
```
Comment (thinkyhead) [inline Marlin/pins_ZRIB_V20.h]: This line does nothing and can be removed.

File: Marlin/pins_ZRIB_V20.h [hunk 9]
```diff
@@ -0,0 +1,13 @@
+#undef SERVO1_PIN 
+#define RAMPS_D6_PIN ZRIB_V20_D6_PIN 
+#define ZRIB_V20_D6_PIN 6
+#define ZRIB_V20_D9_PIN
+#define RAMPS_D9_PIN ZRIB_V20_D6_PIN
+#undef ORIG_E0_AUTO_FAN_PIN
+#undef ORIG_E2_AUTO_FAN_PIN
+#undef ORIG_E3_AUTO_FAN_PIN
```
Comment (thinkyhead) [inline Marlin/pins_ZRIB_V20.h]: These three `#undef` lines also do nothing and can be removed.

File: Marlin/pins_ZRIB_V20.h [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+#define ORIG_E0_AUTO_FAN_PIN ZRIB_V20_D9_PIN
+#define ORIG_E1_AUTO_FAN_PIN ZRIB_V20_D9_PIN
+#define ORIG_E2_AUTO_FAN_PIN ZRIB_V20_D9_PIN
+#define ORIG_E3_AUTO_FAN_PIN ZRIB_V20_D9_PIN
+#include "pins_MKS_13.h"
```
Comment (thinkyhead) [inline Marlin/pins_ZRIB_V20.h]: A blank line above and below this `#include` line would make it better.

Comment (thinkyhead) [reply Marlin/pins_ZRIB_V20.h]: Add a header to the top of this pins file (copy and paste the header from `pins_MELZI.h`, for example).

File: Marlin/boards.h [hunk 11]
```diff
@@ -94,7 +94,7 @@
 #define BOARD_BAM_DICE          401  // 2PrintBeta BAM&DICE with STK drivers
 #define BOARD_BAM_DICE_DUE      402  // 2PrintBeta BAM&DICE Due with STK drivers
 #define BOARD_BQ_ZUM_MEGA_3D    503  // bq ZUM Mega 3D
-
+#define BOARD_ZRIB_V20          504  // zrib V2.0 control board (Chinese knock off RAMPS replica)
```
Comment (thinkyhead) [inline Marlin/boards.h]: Add a blank line below this one to separate it from the definition of `MB` below.

**Review by thinkyhead — CHANGES_REQUESTED** [review]
Overall it looks simple enough. Please make the requested changes and it can be merged soon.

**Comment by FATALHERMIT** [comment]
i'll try to work on it tomorrow

**Comment by FATALHERMIT** [comment]
did the changes but had to start using E0_AUTO_FAN_PIN since it wouldnt work when i used ORIG_E0_AUTOFAN_PIN even pin debugging wasn't recognizing the D9 pin as an autofan when i used it. 
is it a bug or do i need to change something in the configuration files?

**Comment by Ma-P** [comment]
IMHO is a nonsense to include pin_MKS_13.h and not directly pins_RAMPS.h, because the only thing which comes from  pin_MKS_13.h is #define MOSFET_D_PIN 7 .
And there is my version of pin_ZRIB_2.h (see  #6063):

#define MOSFET_D_PIN 6  //FAN_1_PIN
#define IS_RAMPS_EFB

#include "pins_RAMPS.h"

#if HOTENDS == 2
  #undef HEATER_1_PIN
  #HEATER_1_PIN 7
#endif

**Comment by FATALHERMIT** [comment]
so that should be it, sorry for the delay, midterms kept me pretty busy. on a side note, i've added you both to the collaborators on the repositories

**Comment by FATALHERMIT** [comment]
@thinkyhead im ready to be inspected again

**Comment by lordgotham** [comment]
hi sorry for my bad english but i'm french and don't understand all
i have a p802QR2 printer (zonestar) with 2 extruder and sensor with a zrib v2.2
i'm looking for a firmware for working my 2 extruder and sensor, i write many times zonestar but nobody help me.

could you explain me if you firmware can help me?
i see zrib v2

thank you

**Comment by FATALHERMIT** [comment]
@lordgotham from i could tell from comparing your model to mine, which is a P802N or P802M, the P802 series seems to be very similar in terms of specs, so i guess you could try the firmware from the support site of site i bought it from [http://www.3dprintertutorial.org/eg-dual-upgrade](url) just be ready to press the emergency stop just in case

**Comment by empakoso** [comment]
Thank you for your work on the ZRIB board.  I too have this board and would like to update it with your code.  Do you have a working copy of the Marlin fw for your EG printer?  I believe I own the same printer you referenced above. http://3dprintertutorial.org/prusa-i3-with-auto-level-and-zrib-card-tutorial

I would like to work out some of 'my' bugs in 1.1.1 using the new volumemetric, filament width sensor, and linear advance features.  However before I do that, I don't like my modifications I made to the motherboard pins.  I prefer a more professionally coded, approach.  

I've also added a diode laser as "T1" and would want to further that modification, but first want to correct my base fw, the motherboard.

I am looking forward to having this merged.

**Comment by thinkyhead** [comment]
Merged at #6932.

**Pull request closed without merging**